core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java

```diff
@@ -155,9 +181,42 @@ public void removedBundle(Bundle bundle, BundleEvent bundleEvent, Iterable<? ext
                 LOG.debug("No BOM found in {} {} {}", bundleIds(bundle));
             }
 
+            if (!passesWhiteAndBlacklists(bundle)) {
+                removeAnyApplications(catalogItems);
+            }
+
             return catalogItems;
         }
 
+        private void removeAnyApplications(Iterable<? extends CatalogItem<?, ?>> catalogItems) {
+
+            final Iterator<? extends CatalogItem<?, ?>> cit = catalogItems.iterator();
+            while (cit.hasNext()) {
+                final CatalogItem<?, ?> item = cit.next();
+                if (TEMPLATE.equals(item.getCatalogItemType())) {
+                    removeFromCatalog(item);
+                    cit.remove();
+                }
+            }
+        }
+
+        private boolean passesWhiteAndBlacklists(Bundle bundle) {
+            final List<String> white = managementContext.getConfig().getConfig(WHITELIST);
+            final List<String> black = managementContext.getConfig().getConfig(BLACKLIST);
+
+            return on(bundle, white) && !on(bundle, black);
+        }
+
+        private boolean on(Bundle bundle, List<String> list) {
+            for (String candidate : list) {
+                final String symbolicName = bundle.getSymbolicName();
+                if (symbolicName.matches(candidate)) {
```

Comment:
I'll trim the value, it's good to be explicit.  It should work fine with just one item.